Makefile

```diff
@@ -167,3 +169,84 @@ endif
+else ifeq ($(OS),Linux)
+	cd server && env GOOS=linux GOARCH=amd64 $(GO) build -gcflags "all=-N -l" -o dist/plugin-linux-amd64;
+else ifeq ($(OS),Windows_NT)
+	cd server && env GOOS=windows GOARCH=amd64 $(GO) build -gcflags "all=-N -l" -o dist/plugin-windows-amd64.exe;
+else
+	$(error make debug depends on uname to return your OS. If it does not return 'Darwin' (meaning OSX), 'Linux', or 'Windows_NT' (all recent versions of Windows), you will need to edit the Makefile for your own OS.)
+endif
+
+	rm -rf dist/
+	mkdir -p dist/$(PLUGIN_ID)/server/dist
+	cp $(MANIFEST_FILE) dist/$(PLUGIN_ID)/
+	cp -r server/dist/* dist/$(PLUGIN_ID)/server/dist/
+	mkdir -p ../mattermost-server/plugins
+	cp -r dist/* ../mattermost-server/plugins/
+
+# webapp-debug builds and deploys a debug version of the plugin's webapp
+.PHONY: webapp-debug
+webapp-debug:
+
+ifneq ($(HAS_WEBAPP),)
+# link the webapp directory
+	mkdir -p ../mattermost-server/plugins/$(PLUGIN_ID)/webapp
+	ln -nfs $(PWD)/webapp/dist ../mattermost-server/plugins/$(PLUGIN_ID)/webapp/dist
+# start an npm watch
+	cd webapp && $(NPM) run run &
+endif
+
+	@echo "\n\n*** After the frontend is compiled, run 'make reset' to reset the plugin.\nRun 'make reset' every time a change is made to force the server to serve the chages in your webapp portion of the plugin.\nRun 'make stop' to stop the npm watcher.\n\n"
+
+# Reset the plugin
+.PHONY: reset
+reset:
+ifeq ($(and $(MM_SERVICESETTINGS_SITEURL),$(MM_ADMIN_USERNAME),$(MM_ADMIN_PASSWORD)),)
+	$(error In order to use make reset, the following environment variables need to be defined: MM_SERVICESETTINGS_SITEURL, MM_ADMIN_USERNAME, MM_ADMIN_PASSWORD)
+endif
+
+# If we were debugging, we have to unattach the delve process or else we can't disable the plugin.
+# NOTE: we are assuming the dlv was listening on port 2346, as in the debug-plugin.sh script.
+	@DELVE_PID=$(shell ps aux | grep "dlv attach.*2346" | grep -v "grep" | awk -F " " '{print $$2}') && \
+	if [ "$$DELVE_PID" -gt 0 ] > /dev/null 2>&1 ; then \
+		echo "Located existing delve process running with PID: $$DELVE_PID. Killing." ; \
+		kill -9 $$DELVE_PID ; \
+	fi
+
+ifneq ($(CURL),)
+	@echo "\nRestarting plugin via API"
+	$(eval TOKEN := $(shell curl -i -X POST $(MM_SERVICESETTINGS_SITEURL)/api/v4/users/login -d '{"login_id": "$(MM_ADMIN_USERNAME)", "password": "$(MM_ADMIN_PASSWORD)"}' | grep -o MMAUTHTOKEN=[0-9a-z]\* | cut -f2 -d'=' 2> /dev/null))
+	@curl -s -H "Authorization: Bearer $(TOKEN)" -X POST $(MM_SERVICESETTINGS_SITEURL)/api/v4/plugins/$(PLUGIN_ID)/disable > /dev/null && \
+		curl -s -H "Authorization: Bearer $(TOKEN)" -X POST $(MM_SERVICESETTINGS_SITEURL)/api/v4/plugins/$(PLUGIN_ID)/enable > /dev/null && \
+		echo "OK." || echo "Sorry, something went wrong. Check that MM_ADMIN_USERNAME and MM_ADMIN_PASSWORD env variables are set correctly."
+else
+	$(error In order to use make reset, you need to have curl installed.)
+endif
+
+# Stop the webpack
+.PHONY: stop
```

Comment:
Instead of `stop` target, can we make `webapp-debug` blocking, and thus terminated with a SIGINT? This has long been a "mistake" in the mattermost-webapp, and something I hope we can eventually fix there as well.

Reply:
Personally I wouldn't want that -- I use the same terminal to run all these commands, with the webpack running in the background. I can see when the webpack is finished and then I can run the commands I need to. If it was blocking it would kind of negate the benefit of the always-on watching (for me at least)

Reply:
I can appreciate why it seems convenient at first, but there are a number of shortcomings:
* Anyone using a terminal editor (e.g. vim) will be spammed with logs from these background processes inside their editor.
* Running more than one background process makes it hard to decouple logs -- especially failure logs -- from a particular process, not to mention just using the console with something randomly deciding to write makes it hard to enter commands.
* We have often fielded questions from new contributors about why they can't start their server, only to realize they had already started it somewhere else but couldn't "find" it since it was in the background.

I'm 5/5 that we don't want to promulgate this behaviour into the sample plugin, especially when you can trivially opt-in with your own `make <target> &`.

As an aside, I'd strongly encourage the use of tmux or a similar equivalent and running your jobs in separate panes within the same window. At least, I find that way more powerful.

Reply:
Strong :+1: for what Jesse said.